src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -121,142 +116,162 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
+                    break;
+            }
+        }
 
-            // Find matching extension methods from source.
-            foreach (var index in syntaxIndices)
+        private static void GetExtensionMethodItemsWorker(
+            int position,
+            SemanticModel semanticModel,
+            ITypeSymbol receiverTypeSymbol,
+            ImmutableArray<IMethodSymbol> matchingMethodSymbols,
+            bool isSymbolFromCurrentCompilation,
+            ArrayBuilder<SerializableImportCompletionItem> builder,
+            Dictionary<INamespaceSymbol, string> stringCache)
+        {
+            foreach (var methodSymbol in matchingMethodSymbols)
             {
-                // Add simple extension methods with matching target type name
-                foreach (var targetTypeName in targetTypeNames)
-                {
-                    var methodInfos = index.SimpleExtensionMethodInfo[targetTypeName];
-                    if (methodInfos.Count == 0)
-                    {
-                        continue;
-                    }
+                var methodSymbolInCurrentCompilation = isSymbolFromCurrentCompilation
+                    ? methodSymbol
+                    : SymbolFinder.FindSimilarSymbols(methodSymbol, semanticModel.Compilation).FirstOrDefault();
 
-                    foreach (var methodInfo in methodInfos)
-                    {
-                        results.Add(methodInfo.FullyQualifiedContainerName, methodInfo.Name);
-                    }
+                if (methodSymbolInCurrentCompilation == null
+                    || !semanticModel.IsAccessible(position, methodSymbolInCurrentCompilation))
+                {
+                    continue;
                 }
 
-                // Add all complex extension methods, we will need to completely rely on symbols to match them.
-                foreach (var methodInfo in index.ComplexExtensionMethodInfo)
+                var reducedMethodSymbol = methodSymbolInCurrentCompilation.ReduceExtensionMethod(receiverTypeSymbol);
+                if (reducedMethodSymbol != null)
                 {
-                    results.Add(methodInfo.FullyQualifiedContainerName, methodInfo.Name);
+                    var symbolKeyData = SymbolKey.CreateString(reducedMethodSymbol);
+                    builder.Add(new SerializableImportCompletionItem(
+                        symbolKeyData,
+                        reducedMethodSymbol.Name,
+                        reducedMethodSymbol.Arity,
+                        reducedMethodSymbol.GetGlyph(),
+                        GetFullyQualifiedNamespaceName(reducedMethodSymbol.ContainingNamespace, stringCache)));
                 }
             }
+        }
 
-            // Find matching extension methods from metadata
-            foreach (var info in symbolInfos)
+        private static string GetFullyQualifiedNamespaceName(INamespaceSymbol symbol, Dictionary<INamespaceSymbol, string> stringCache)
```

Comment:
are there any threading concerns here?  i.e. is ther eany concurrency here?

Reply:
Right now it's single threaded, although we might want to try making it parallel for potential perf gains.